crates/subspace-networking/src/utils/circular_buffer.rs

```diff
@@ -0,0 +1,281 @@
+    ///
+    /// # Returns
+    /// - `Ok(_)`: If the element add was successful.
+    ///     - `Some(T)`: If adding an element resulted in the removal of an
+    ///         existing one (in the case of a circular buffer, for instance)
+    ///     - `None`: Adding an element did not return any value
+    /// - `Error`: If the element add was unsuccessful
+    ///
+    /// # Errors
+    /// Attempting to add an element to a full queue that does not allow for
+    /// overflow will return an error.
+    fn add(&mut self, val: T) -> Result<Option<T>, &str>;
+
+    /// Removes an element from the queue and returns it
+    ///
+    /// For queues with default values, removing an element will add a new
+    /// default value into the queue.
+    ///
+    /// # Returns
+    /// - `Ok(T)`: The oldest element in the queue
+    /// - `Error`
+    ///
+    /// # Errors
+    /// Returns an error if an attempt is made to remove an element from
+    /// an empty queue
+    fn remove(&mut self) -> Result<T, &str>;
+
+    /// Peek at the head of the queue
+    ///
+    /// # Returns
+    /// - `Ok(T)`: The next element scheduled for removal from the queue
+    /// - `Error`
+    ///
+    /// # Errors
+    /// Returns an error if an attempt is made to peek into an empty queue
+    fn peek(&self) -> Result<T, &str>;
+
+    /// Gets the size of the queue
+    ///
+    /// # Returns
+    /// The number of elements in the queue. Note, this _includes_ default
+    /// values when specified, which means that the `size` of a queue with
+    /// default values should always be equal to its `capacity`
+    fn size(&self) -> usize;
+
+    /// Remove all entries of a given value.
+    fn remove_value(&mut self, val: &T);
+}
+
+/// Represents a FIFO `CircularBuffer<T>` data structure.
+///
+/// This structure is a limited capacity queue, with optional provisions
+/// for default values. Under normal circumstances, the `size` of the
+/// queue grows until it reaches its `capacity`, at which point any
+/// further additions push out its oldest member.
+///
+/// If default values are specified, then the `size` of the queue
+/// is always equal to its `capacity`, with empty slots occupied by the
+/// specified default value.
+///
+/// # Type parameters
+/// - `T`: Any type that implements the `Clone` trait.
+///
+/// # Examples
+///
+/// ```
+/// # use queues::*;
+/// # fn main() {
+/// let mut cbuf = CircularBuffer::<isize>::new(3);
+/// let mut cbuf_def = CircularBuffer::with_default(3, 0isize);
+///
+/// // Check sizes
+/// assert_eq!(cbuf.size(), 0);
+/// assert_eq!(cbuf_def.size(), 3);
+///
+/// // Add elements
+/// cbuf.add(6);
+/// cbuf_def.add(7);
+///
+/// // Peek at the next element scheduled for removal
+/// assert_eq!(cbuf.peek().unwrap(), 6);
+/// assert_eq!(cbuf_def.peek().unwrap(), 0);
+/// # }
+/// ```
+#[derive(Debug)]
+pub struct CircularBuffer<T: Clone> {
+    queue: Vec<T>,
+    capacity: usize,
+    default_value: Option<T>,
+}
```

Comment:
What about using off-the-shelf https://doc.rust-lang.org/stable/std/collections/struct.VecDeque.html instead of rolling essentially the same behavior, but way more efficient removal from the beginning of the list?

Reply:
You can check `queue.len() == queue.capacity()` and if so then remove oldest value before inserting a new one. That way you get bounded size.

Reply:
Correct, and this will be the creation of the CircularBuffer features. It's not a ready-to-use type. 

Anyway, I added this issue and will fix it soon: https://github.com/subspace/subspace/issues/903